### PR TITLE
[project-base] parameter data fixture refactoring

### PR DESCRIPTION
**Pull request by AndrejBlaho**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Defining product parameter values in data fixtures was overly complicated accounting also for ad hoc parameter creation. Parameters are now created solely in parameter data fixtures and process of parameter value definition is generally simplified and more readable. 
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

----
:globe_with_meridians: Live Preview:
  - https://mg-uuid-pool.odin.shopsys.cloud
  - https://cz.mg-uuid-pool.odin.shopsys.cloud

### CODE REVIEW

File: project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php [hunk 1]
```diff
@@ -44,26 +116,210 @@ public function __construct(
...
+     * @return array<string, string|\App\DataFixtures\Demo\ParameterDataFixtureData>
+     */
+    private function getParameterData(string $locale): array
+    {
+        if (isset(self::$parameterNameCacheByLocale[$locale])) {
+            return self::$parameterNameCacheByLocale[$locale];
+        }
+
+        /** @var \App\Model\Product\Parameter\ParameterGroup $mainInformationParameterGroup */
+        $mainInformationParameterGroup = $this->getReference(ParameterGroupDataFixture::PARAM_GROUP_MAIN_INFORMATION);
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php]: thanks to https://github.com/shopsys/shopsys/pull/3070, is now possible to specify a type directly in the method call, can you please rewrite it in all relevant files in this PR?

File: project-base/app/src/DataFixtures/Demo/UuidFactory.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+declare(strict_types=1);
+
+namespace App\DataFixtures\Demo;
+
+use Override;
+use Ramsey\Uuid\UuidFactory as BaseUuidFactory;
+use Ramsey\Uuid\UuidInterface;
+
+class UuidFactory extends BaseUuidFactory
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/UuidFactory.php]: I don't think this class is still relevant... The uuid pools were refactored in https://github.com/shopsys/shopsys/pull/3075

File: project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php [hunk 3]
```diff
@@ -29,12 +31,85 @@ class ParameterDataFixture extends AbstractReferenceFixture implements Dependent
     public const PARAMETER_PREFIX = 'parameter_';
     public const PARAMETER_SLIDER_WARRANTY = 'parameter_slider_warranty';
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php]: those two constants should be removed as the related reference is not set

File: project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php [hunk 4]
```diff
@@ -44,26 +119,213 @@ public function __construct(
...
         $parameterColorNamesByLocale = [];
         $parameterMaterialNamesByLocale = [];
 
         foreach ($this->domain->getAllLocales() as $locale) {
             $parameterColorNamesByLocale[$locale] = t('Color', [], Translator::DATA_FIXTURES_TRANSLATION_DOMAIN, $locale);
             $parameterMaterialNamesByLocale[$locale] = t('Material', [], Translator::DATA_FIXTURES_TRANSLATION_DOMAIN, $locale);
         }
-        $parameterColor = $this->createParameter(
+
+        $this->createParameter(
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php]: these two parameters (color, material) should be already created from the previous foreach, or?

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php (hunk 1 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php]: sure, filled it in everywhere

File: project-base/app/src/DataFixtures/Demo/UuidFactory.php (hunk 2 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/UuidFactory.php]: indeed it is not, removed it

File: project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php (hunk 3 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php]: removed as well

File: project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php (hunk 4 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php]: yes, you are right, now is everything moved to its correct place

File: project-base/app/src/DataFixtures/Demo/ReadyCategorySeoDataFixture.php [hunk 5]
```diff
@@ -245,9 +245,9 @@ public function load(ObjectManager $manager)
             t('meta description of TV, audio plasma with HDMI seo category', [], Translator::DATA_FIXTURES_TRANSLATION_DOMAIN, $firstDomainLocale),
         );
 
-        $categoryPc = $this->getReference(CategoryDataFixture::CATEGORY_PC, Category::class);
-        $newFlag = $this->getReference(FlagDataFixture::FLAG_PRODUCT_NEW, Flag::class);
-        $usbParameter = $this->getReference(ParameterDataFixture::PARAMETER_PREFIX . t('USB', [], Translator::DATA_FIXTURES_TRANSLATION_DOMAIN, $firstDomainLocale), Parameter::class);
+        $categoryPc = $this->getReference(CategoryDataFixture::CATEGORY_PC);
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/ReadyCategorySeoDataFixture.php]:
```suggestion
        $categoryPc = $this->getReference(CategoryDataFixture::CATEGORY_PC, Category::class);
```

File: project-base/app/src/DataFixtures/Demo/ReadyCategorySeoDataFixture.php [hunk 6]
```diff
@@ -245,9 +245,9 @@ public function load(ObjectManager $manager)
...
+        $newFlag = $this->getReference(FlagDataFixture::FLAG_PRODUCT_NEW);
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/ReadyCategorySeoDataFixture.php]:
```suggestion
        $newFlag = $this->getReference(FlagDataFixture::FLAG_PRODUCT_NEW, Flag::class);
```

File: project-base/app/src/DataFixtures/Demo/ParameterColorValueDataFixture.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            $parameterValueData = $this->parameterValueDataFactory->create();
+            $parameterValueData->locale = $locale;
+            $parameterValueData->text = t('black', [], Translator::DATA_FIXTURES_TRANSLATION_DOMAIN, $locale);
+            $parameterValueData->rgbHex = '#000000';
+            $this->parameterRepository->findOrCreateParameterValueByParameterValueData($parameterValueData);
+
+            $parameterValueData = $this->parameterValueDataFactory->create();
+            $parameterValueData->locale = $locale;
+            $parameterValueData->text = t('white', [], Translator::DATA_FIXTURES_TRANSLATION_DOMAIN, $locale);
+            $parameterValueData->rgbHex = '#FFFFFF';
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/ParameterColorValueDataFixture.php]: nitpick:

can you please use lowercased letters here? Now it's inconsistent with the next rgbhex

File: project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php [hunk 8]
```diff
@@ -4,123 +4,297 @@
...
     }
 
     /**
-     * @param \Doctrine\Persistence\ObjectManager $manager
+     * @param string $referenceName
+     * @param string $locale
+     * @return string
      */
-    public function load(ObjectManager $manager)
+    public function getParameterNameByReferenceName(string $referenceName, string $locale): string
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php]: this method may be private, or?

**Review by grossmannmartin — CHANGES_REQUESTED** [review]
nice job, can you please check my comments?

File: project-base/app/src/DataFixtures/Demo/ReadyCategorySeoDataFixture.php (hunk 5 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/ReadyCategorySeoDataFixture.php]: added

File: project-base/app/src/DataFixtures/Demo/ReadyCategorySeoDataFixture.php (hunk 6 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/ReadyCategorySeoDataFixture.php]: added here and also to other places where missing

File: project-base/app/src/DataFixtures/Demo/ParameterColorValueDataFixture.php (hunk 7 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/ParameterColorValueDataFixture.php]: sure, now it is unified

File: project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php (hunk 8 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/ParameterDataFixture.php]: yes it should, fixed it

**Review by grossmannmartin — APPROVED** [review]

File: project-base/app/src/DataFixtures/Demo/DataObjects/ParameterDataFixtureData.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\DataFixtures\Demo\DataObjects;
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/DataObjects/ParameterDataFixtureData.php]: A namespace is usually named in singular, can you please rename it to `DataObject`?

**Review by grossmannmartin — APPROVED** [review]

File: project-base/app/src/DataFixtures/Demo/DataObjects/ParameterDataFixtureData.php (hunk 9 above)
Comment (AndrejBlaho) [reply project-base/app/src/DataFixtures/Demo/DataObjects/ParameterDataFixtureData.php]: Sure. Changed it to singular form.

**Pull request merged**